contracts/FastBreak.cdc

```diff
@@ -0,0 +1,682 @@
+        fastBreakRunID: String,
+        isPublic: Bool,
+        submissionDeadline: UInt64,
+        numPlayers: UInt64
+    )
+    pub event FastBreakGameStatusChange(id: String, newStatus: String)
+    pub event FastBreakNFTBurned(id: UInt64, serialNumber: UInt64)
+    pub event FastBreakNFTMinted(
+        id: UInt64,
+        fastBreakGameID: String,
+        serialNumber: UInt64,
+        mintingDate: UInt64,
+        topShots: [UInt64],
+        mintedTo: Address
+    )
+    pub event FastBreakSubmissionSent(
+        wallet: Address,
+        submittedAt: UInt64,
+        fastBreakGameID: String,
+        topShots: [UInt64]
+    )
+    pub event FastBreakGameWinner(
+        wallet: Address?,
+        submittedAt: UInt64,
+        fastBreakGameID: String,
+        topShots: [UInt64]
+    )
+    pub event FastBreakGameStatAdded(
+        fastBreakGameID: String,
+        name: String,
+        type: String,
+        valueNeeded: UInt64
+    )
+
+    /// Named Paths
+    ///
+    pub let CollectionStoragePath:  StoragePath
+    pub let CollectionPublicPath:   PublicPath
+    pub let OracleStoragePath:       StoragePath
+    pub let OraclePrivatePath:      PrivatePath
+
+    /// Contract variables
+    ///
+    pub var totalSupply:        UInt64
+
+    /// Metadata Dictionaries
+    ///
+    access(self) let fastBreakRunByID:        {String: FastBreakRun}
+    access(self) let fastBreakGameByID:           {String: FastBreakGame}
+
+    /// A top-level Fast Break Run, the container for Fast Break Games
+    ///
+    pub struct FastBreakRun {
+        pub let id: String
+        pub let name: String
+        pub var status: String
+        pub let runStart: UInt64
+        pub let runEnd: UInt64
+        pub let leaderboard: {Address: UInt64}
+        pub let fatigueModeOn: Bool
+
+        init (id: String, name: String, runStart: UInt64, runEnd: UInt64, fatigueModeOn: Bool) {
+            if let fastBreakRun = FastBreak.fastBreakRunByID[id] {
+                self.id = fastBreakRun.id
+                self.name = fastBreakRun.name
+                self.status = fastBreakRun.status
+                self.runStart = fastBreakRun.runStart
+                self.runEnd = fastBreakRun.runEnd
+                self.leaderboard = fastBreakRun.leaderboard
+                self.fatigueModeOn = fastBreakRun.fatigueModeOn
+            } else {
+                self.id = id
+                self.name = name
+                self.status = "FAST_BREAK_RUN_OPEN"
+                self.runStart = runStart
+                self.runEnd = runEnd
+                self.leaderboard = {}
+                self.fatigueModeOn = fatigueModeOn
+            }
+        }
+
+        /// Update status of the Fast Break Run
+        ///
+        access(contract) fun updateStatus(status: String) { self.status = status }
+
+        /// Write a new win to the Fast Break Run leaderboard
+        ///
+        access(contract) fun incrementLeaderboardWins(wallet: Address) {
```

Comment:
probably should use a different name than `wallet`, like `account`, or something. It might also be better to tie stats to resource IDs instead of accounts too because that is more accurate to the cadence programming model

Reply:
It also would allow people to trade and sell their fastbreak player records since the records are tied to NFTs instead of accounts, so it is another feature that takes advantage of web3 unique features

Reply:
I have renamed wallet to account. Good call.

I gave the concept of using a resourceId a lot of thought and it has advantages like you describe. I never considered that originally, so it was fun to think through. The reason I chose a flow account address here is two-fold. First, I made the off-chain version first and I need to tie non-custodial submissions to something I know about off-chain which is flowAddress. Next, this game token is "soul-bound" lite in that it records the address of who won it, but does not restrict ownership like a true soul-bound token would. There are a lot of opinions on this and have run this concept up to Dete for a related project and this is his preferred "trophy" implementation. Record the address of the winner and keep the token an NFT. People building on this can choose to use or discard this additional data of the token.

I do dig the concept of resourceIds here but our systems are built around flow address and am going to go this route to make integration easier.